Quantized leaky relu

z-a-f opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#33004 Quantized leaky relu**

Differential Revision: [D19740193](https://our.internmc.facebook.com/intern/diff/D19740193)

Commit:
Quantized leaky relu

Commit:
Update on "Quantized leaky relu"


Differential Revision: [D19740193](https://our.internmc.facebook.com/intern/diff/D19740193)

jamesr66a commented on aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:
```diff
@@ -193,6 +193,51 @@ void qrelu6_kernel(const Tensor& qx, Tensor& qy) {
+    using Vec = Vec256<float>;  // Naïve implementation uses dequant/quant loop.
+    using qVec = Vec256<scalar_t>;
+    Vec zero_vec = Vec((float)(0.0f));
+    Vec one_vec = Vec((float)(1.0f));
```
Why do we need explicit casts on float literals?

z-a-f replied:
We don't -- just a habit :)

jamesr66a commented on aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:
```diff
@@ -193,6 +193,51 @@ void qrelu6_kernel(const Tensor& qx, Tensor& qy) {
+
+  int64_t o_zp = out.q_zero_point();
+  float o_scale = out.q_scale();
+  float o_inv_scale = 1.0 / o_scale;
```
1.0f

jamesr66a commented on aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:
```diff
@@ -193,6 +193,51 @@ void qrelu6_kernel(const Tensor& qx, Tensor& qy) {
+                                              i_scale_zp_neg_premul_vec);
+          for (int idx = 0; idx < dx_vec_vec.size(); ++idx) {
+            const auto dx_vec = dx_vec_vec[idx];
+            const auto r = Vec::blendv(negval_vec, one_vec, dx_vec > zero_vec);
```
Can we either name this something more semantic than `r` or place a mathematical formula in a comment that people can refer to to see what `r` is?

jamesr66a commented on test/test_quantized.py:
```diff
@@ -148,6 +148,27 @@ def test_qrelu6(self, X):
+                       qparams=hu.qparams()))
+    def test_qrelu_leaky(self, X):
+        X, (scale, zero_point, torch_type) = X
+        alpha = 0.1
```
Should we test different values for this? Though i like the choice of `0.1` for its pathological floating point representation :)

jamesr66a reviewed (approved):
LG, small comments

Commit:
Update on "Quantized leaky relu"


Differential Revision: [D19740193](https://our.internmc.facebook.com/intern/diff/D19740193)
